src/poly/multiopen/prover.rs

```diff
@@ -0,0 +1,191 @@
+use super::super::{
+    commitment::{self, Blind, Params},
+    Coeff, Error, Polynomial,
+};
+use super::{construct_intermediate_sets, Proof, ProverQuery, Query};
+
+use crate::arithmetic::{
+    eval_polynomial, get_challenge_scalar, kate_division, lagrange_interpolate, Challenge, Curve,
+    CurveAffine, Field,
+};
+use crate::plonk::hash_point;
+use crate::transcript::Hasher;
+use std::marker::PhantomData;
+
+#[derive(Debug, Clone)]
+struct CommitmentData<C: CurveAffine> {
+    set_index: usize,
+    blind: Blind<C::Scalar>,
+    point_indices: Vec<usize>,
+    evals: Vec<C::Scalar>,
+}
+
+impl<C: CurveAffine> Proof<C> {
+    /// Create a multi-opening proof
+    pub fn create<'a, I, HBase: Hasher<C::Base>, HScalar: Hasher<C::Scalar>>(
+        params: &Params<C>,
+        transcript: &mut HBase,
+        transcript_scalar: &mut HScalar,
+        queries: I,
+    ) -> Result<Self, Error>
+    where
+        I: IntoIterator<Item = ProverQuery<'a, C>> + Clone,
+    {
+        let x_4: C::Scalar = get_challenge_scalar(Challenge(transcript.squeeze().get_lower_128()));
+        let x_5: C::Scalar = get_challenge_scalar(Challenge(transcript.squeeze().get_lower_128()));
+
+        let (poly_map, point_sets) = construct_intermediate_sets(queries);
+
+        // Collapse openings at same point sets together into single openings using
+        // x_4 challenge.
+        let mut q_polys: Vec<Option<Polynomial<C::Scalar, Coeff>>> = vec![None; point_sets.len()];
+        let mut q_blinds = vec![Blind(C::Scalar::zero()); point_sets.len()];
+
+        // A vec of vecs of evals. The outer vec corresponds to the point set,
+        // while the inner vec corresponds to the points in a particular set.
+        let mut q_eval_sets: Vec<Vec<_>> = vec![Vec::new(); point_sets.len()];
+        for (set_idx, point_set) in point_sets.iter().enumerate() {
+            q_eval_sets[set_idx] = vec![C::Scalar::zero(); point_set.len()];
+        }
```

Comment:
Simpler would be to `Vec::with_capacity(point_sets.len())` here and not do an enumeration in the loop below; just push to the vector.

Reply:
the difficulty i'm facing is that `accumulate` has to index into `q_eval_sets` at a certain `set_idx` for each new polynomial, and it's not guaranteed that these `set_idx`s will be in ascending order.

Reply:
In this particular code, we're just initializing the vectors in ascending order.

Reply:
oh! i misread the as a larger refactor than actually suggested.